tests/gen/gen_predefined.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package gen
+
+import (
+	"strings"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+)
+
+// GeneratePredefinedFromSDL generates documents for GraphQL SDL from a predefined list
+// of docs that might include nested docs.
+// The SDL is parsed to get the list of fields, and the docs
+// are created with the fields parsed from the SDL.
+// This allows us to have only one large list of docs with predefined
+// fields, and create SDLs with different fields from it.
+func GeneratePredefinedFromSDL(gqlSDL string, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+// GeneratePredefined generates documents from a predefined list
+// of docs that might include nested docs.
+func GeneratePredefined(defs []client.CollectionDefinition, docsList DocsList) ([]GeneratedDoc, error) {
+	resultDocs := make([]GeneratedDoc, 0, len(docsList.Docs))
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, col := range defs {
+		typeDefs[col.Description.Name] = col
+	}
+	generator := docGenerator{types: typeDefs}
+	for _, doc := range docsList.Docs {
+		docs, err := generator.generateRelatedDocs(doc, docsList.ColName)
+		if err != nil {
+			return nil, err
+		}
+		resultDocs = append(resultDocs, docs...)
+	}
+	return resultDocs, nil
+}
+
+type docGenerator struct {
+	types map[string]client.CollectionDefinition
+}
+
+// toRequestedDoc removes the fields that are not in the SDL of the collection.
+func toRequestedDoc(doc map[string]any, typeDef *client.CollectionDefinition) map[string]any {
+	result := make(map[string]any)
+	for _, field := range typeDef.Schema.Fields {
+		if field.IsRelation() || field.Name == request.KeyFieldName {
+			continue
+		}
+		result[field.Name] = doc[field.Name]
+	}
+	for name, val := range doc {
+		if strings.HasSuffix(name, request.RelatedObjectID) {
+			result[name] = val
+		}
+	}
+	return result
+}
+
+// generatePrimary generates primary docs for the given secondary doc.
```

Comment:
todo: I think this documentation needs to highlight that it will essentially mutate the given `docMap` document, appending the keys of any generated secondary documents and then returning a `client.document` that has additional field values to the input `docMap`, and as a result, a different dockey.

Reply:
extended doc